### PR TITLE
Fix various issues in the test makefiles

### PR DESCRIPTION
**Pull request by torarnv**
- Fix broken glob dependency from test bin targets
- Ensure sharness test (clean, build, test, aggregate) is run sequentially
- Ensure we don't remake binaries due to missing IPFS-BUILD-OPTIONS

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -34,27 +27,35 @@ nofuse:
...
+test_go: build
+	go test $(GO_TEST_FLAGS) ./...
 
-test_go_race:
-	$(go_test) ./... -race
+test_sharness_short: test_sharness
 
-test_sharness_short:
-	cd test/sharness/ && make
+test_sharness_expensive: SHARNESS_MAKE_FLAGS += TEST_EXPENSIVE=1
```
Comment (jbenet) [inline Makefile]: i prefer explicitness in these targets. adding flags like that is errorprone in my opinion.

**Comment by torarnv** [comment]
@chriscool :pager:

File: test/sharness/Makefile [hunk 2]
```diff
@@ -14,6 +14,9 @@ IPFS_ROOT = ../..
 # User might want to override those on the command line
 GOFLAGS =
 
+# The targets in this makefile depend on being run sequentially
+.NOTPARALLEL:
```
Comment (jbenet) [inline test/sharness/Makefile]: :+1: i would love to move towards being able to run them in parallel. wonder what git does + how it would manage the sharness output. cc @chriscool

File: test/Makefile [hunk 3]
```diff
@@ -18,21 +18,14 @@ clean:
...
-bin/pollEndpoint: $(POLLENDPOINT_SRC)/*.go IPFS-BUILD-OPTIONS
-	@echo "*** installing $@ ***"
-	go build $(GOFLAGS) -o bin/pollEndpoint $(POLLENDPOINT_SRC)
+bin/random: BIN_SOURCES = $(RANDOM_SRC)/random
+bin/multihash: BIN_SOURCES = $(MULTIHASH_SRC)/multihash
+bin/ipfs: BIN_SOURCES = $(IPFS_CMD)
+bin/pollEndpoint: BIN_SOURCES = $(POLLENDPOINT_SRC)
+
+bin/%: FORCE IPFS-BUILD-OPTIONS
+	@echo "*** building and installing test local version of '$*' binary ***"
```
Comment (jbenet) [inline test/Makefile]: i think the shorter message is fine. 

if the goal is to document what's going on here, maybe we can add a note to the readme or the makefile itself.

File: Makefile [hunk 4]
```diff
@@ -34,27 +27,35 @@ nofuse:
...
 	cd test/3nodetest && make
 
-test_go_short:
-	$(go_test) -test.short ./...
+ifeq ($(TEST_NO_FUSE),1)
+GO_TEST_FLAGS = -tags nofuse
+endif
+
+test_go_short: GO_TEST_FLAGS += -test.short
+test_go_short: test_go
```
Comment (jbenet) [inline Makefile]: this is harder to understand and modify. :-1:

**Comment by jbenet** [comment]
copying comments from the commit:

- We dont want to have to rebuild on every test run-- this is unnecessarily slow.
- if `**/*.go` doesnt work with gnu make, maybe we can use something like
    ```
    $(shell find $(IPFS_ROOT) | egrep ".go$")
    ```
  (not sure if this works)

File: Makefile (hunk 1 above)
Comment (torarnv) [inline Makefile]: Fine by me, not married to  7acee02 and   086f038, those were just drive-by changes on the way to the other two. Personally I think duplicated code is more error prone, as things are changed one place but missed in another, and as a new contributed I find logic easier to understand when I don't have to scan 4 instances of it to see if there's any difference in them.

**Comment by jbenet** [comment]
i dont feel strongly about most comments above. only the rebuilds.

let's hear @chriscool's opinions?

**Comment by jbenet** [comment]
Also, good find on both the parallel and gnu make bugs.

File: test/sharness/Makefile (hunk 2 above)
Comment (torarnv) [inline test/sharness/Makefile]: We may get away with adding explicit deps in the makefile. I wasn't sure if the tests themselves could be run in parallell. The main problem that spawned this is that with `make -j4` eg, the tests will start running before the clean and the deps targets have been run, which resulted in the missing ipfs binary error. Let me try that real quick.

File: test/Makefile (hunk 3 above)
Comment (torarnv) [inline test/Makefile]: The goal was based on my subjective preference of not having to scan N instances of a logic (makefile recipe, function, etc) to see if there's any difference in them, when learning new code.

But I'm fine with abandoning this hunk of the PR.

Comment (chriscool) [inline test/Makefile]: I am ok with 086f038b3f920e901f7be94e1fff278aa9858dfe if the previous commit is fixed, though I agree with @jbenet that a shorter message is better.

File: test/sharness/Makefile (hunk 2 above)
Comment (chriscool) [inline test/sharness/Makefile]: To run tests in parallel I think we need something like:

```
all:
	make clean
	make deps
	make $(T)
	make aggregate
```

and then use `make -j 4` for example.

**Comment by torarnv** [comment]
@chriscool 

>  the IPFS-BUILD-OPTIONS target depends on FORCE but that doesn't mean that we are re-building the binaries on every invocation of make.

Actually, as long as `$(GOFLAGS)` is never set, we never write `IPFS-BUILD-OPTIONS`, so the dependency will always fail, and we'll re-make the binaries every time:

```
{vegas:~/dev/ipfs/src/github.com/ipfs/go-ipfs/test} $ rm -f IPFS-BUILD-OPTIONS
{vegas:~/dev/ipfs/src/github.com/ipfs/go-ipfs/test} $ make bin/ipfs --debug
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
Reading makefiles...
Updating goal targets....
   File `IPFS-BUILD-OPTIONS' does not exist.
     File `FORCE' does not exist.
    Must remake target `FORCE'.
    Successfully remade target file `FORCE'.
  Must remake target `IPFS-BUILD-OPTIONS'.
 Prerequisite `IPFS-BUILD-OPTIONS' of target `bin/ipfs' does not exist.
Must remake target `bin/ipfs'.
*** installing bin/ipfs ***
go build  -o bin/ipfs ../cmd/ipfs
{vegas:~/dev/ipfs/src/github.com/ipfs/go-ipfs/test} $ make bin/ipfs --debug
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
Reading makefiles...
Updating goal targets....
   File `IPFS-BUILD-OPTIONS' does not exist.
     File `FORCE' does not exist.
    Must remake target `FORCE'.
    Successfully remade target file `FORCE'.
  Must remake target `IPFS-BUILD-OPTIONS'.
 Prerequisite `IPFS-BUILD-OPTIONS' of target `bin/ipfs' does not exist.
Must remake target `bin/ipfs'.
*** installing bin/ipfs ***
go build  -o bin/ipfs ../cmd/ipfs
```

I'll fix that in a separate commit.

File: test/sharness/Makefile (hunk 2 above)
Comment (torarnv) [inline test/sharness/Makefile]: b8bc885 adds the proper deps

**Comment by torarnv** [comment]
Okey, ready for review now, removed the subjective refactoring and focused on fixing the issues.

**Comment by chriscool** [comment]
Ok, it LGTM, thanks @torarnv !

**Comment by torarnv** [comment]
Thanks! Happy to help! :smile:

File: test/sharness/Makefile (hunk 2 above)
Comment (chriscool) [inline test/sharness/Makefile]: @jbenet about Git I had a look and yeah some work has been done a long time ago so that the tests can run in parallel (see https://github.com/git/git/commit/abc5d372ec242fc654dc6780df6ea3d63dc72f2f).

And there are always some flags or options or other stuff in the IPFS-BUILD-OPTIONS like files, so that there is no problem with empty files.

**Comment by jbenet** [comment]
Thanks @torarnv !
